File: tools/peers/julia_ihw.jl
#!/usr/bin/env julia

using DelimitedFiles

function get_arg(name)
    index = findfirst(isequal(name), ARGS)
    if index === nothing || index == length(ARGS)
        return nothing
    end
    return ARGS[index + 1]
end

function fail(message, status)
    println(stderr, message)
    exit(status)
end

pvalues_path = get_arg("--pvalues")
covariates_path = get_arg("--covariates")
alpha_text = get_arg("--alpha")
nbins_text = get_arg("--nbins")
nfolds_text = get_arg("--nfolds")
lambda_policy = get_arg("--lambda-policy")
seed_text = get_arg("--seed")
output_prefix = get_arg("--output-prefix")
if any(isnothing, (pvalues_path, covariates_path, alpha_text, nbins_text, nfolds_text, lambda_policy, seed_text, output_prefix))
    fail("Julia IHW adapter arguments are incomplete", 1)
end

try
    using IndependentHypothesisWeighting
catch error
    fail("IndependentHypothesisWeighting.jl is not installed: $(error)", 3)
end

pvalues = vec(Float64.(readdlm(pvalues_path)))
covariates = vec(Float64.(readdlm(covariates_path)))
if isempty(pvalues) || length(pvalues) != length(covariates)
    fail("Julia IHW adapter received invalid input lengths", 1)
end

fit_function = if isdefined(IndependentHypothesisWeighting, :adjust_ihw)
    getfield(IndependentHypothesisWeighting, :adjust_ihw)
elseif isdefined(IndependentHypothesisWeighting, :ihw)
    getfield(IndependentHypothesisWeighting, :ihw)
else
    fail("IndependentHypothesisWeighting.jl exposes no recognized fitting function", 3)
end

lambda_value = lambda_policy == "auto" ? :auto : Inf
fit = try
    fit_function(
        pvalues,
        covariates;
        alpha = parse(Float64, alpha_text),
        nbins = parse(Int, nbins_text),
        nfolds = parse(Int, nfolds_text),
        lambdas = lambda_value,
        seed = parse(Int, seed_text)
    )
catch error
    fail("IndependentHypothesisWeighting.jl call failed: $(error)", 1)
end

if !(hasproperty(fit, :adj_pvalues) && hasproperty(fit, :weights))
    fail("IndependentHypothesisWeighting.jl result has no recognized output fields", 3)
end
adjusted = Float64.(getproperty(fit, :adj_pvalues))
weights = Float64.(getproperty(fit, :weights))
rejections = sum(adjusted .<= parse(Float64, alpha_text))
writedlm(string(output_prefix, ".adj.txt"), adjusted)
writedlm(string(output_prefix, ".weights.txt"), weights)
writedlm(string(output_prefix, ".rejections.txt"), [rejections])

File: tools/peers/julia_ihw.py
"""Adapt the optional Julia IHW package to the peer result contract."""

from __future__ import annotations

import shutil
import subprocess
import sys
import tempfile
from pathlib import Path

import numpy as np

ROOT = Path(__file__).resolve().parents[2]
sys.path.insert(0, str(ROOT))

from tools.data_contract import PeerInput
from tools.peers.runner import (
    AdapterUnavailable,
    FitResult,
    RunConfig,
    adapter_main,
)

JULIA_SCRIPT = Path(__file__).with_name("julia_ihw.jl")
IMPLEMENTATION_VERSION = "IndependentHypothesisWeighting.jl-adapter-runtime-version"


def fit(peer_input: PeerInput, config: RunConfig) -> FitResult:
    """Run the preliminary Julia adapter when its runtime and package exist."""

    julia = shutil.which("julia")
    if julia is None:
        raise AdapterUnavailable("julia is not installed")
    if peer_input.groups is not None or peer_input.folds is not None:
        raise AdapterUnavailable(
            "the Julia adapter does not accept frozen groups and folds"
        )
    nfolds = 5 if config.nfolds is None else config.nfolds
    nbins = (
        max(1, min(40, peer_input.size // 1500))
        if config.nbins == "auto"
        else config.nbins
    )
    if not isinstance(nbins, int):
        raise TypeError("Julia adapter could not resolve nbins")
    with tempfile.TemporaryDirectory(prefix="ihwkit_julia_peer_") as temporary:
        directory = Path(temporary)
        pvalues_path = directory / "pvalues.txt"
        covariates_path = directory / "covariates.txt"
        output_prefix = directory / "result"
        np.savetxt(pvalues_path, peer_input.pvalues, fmt="%.17g")
        np.savetxt(covariates_path, peer_input.covariates, fmt="%.17g")
        command = [
            julia,
            "--startup-file=no",
            str(JULIA_SCRIPT),
            "--pvalues",
            str(pvalues_path),
            "--covariates",
            str(covariates_path),
            "--alpha",
            str(config.alpha),
            "--nbins",
            str(nbins),
            "--nfolds",
            str(nfolds),
            "--lambda-policy",
            config.lambda_policy,
            "--seed",
            str(config.seed),
            "--output-prefix",
            str(output_prefix),
        ]
        completed = subprocess.run(
            command,
            cwd=ROOT,
            capture_output=True,
            text=True,
            check=False,
        )
        detail = completed.stderr.strip() or completed.stdout.strip()
        if completed.returncode == 3:
            raise AdapterUnavailable(detail or "Julia IHW is unavailable")
        if completed.returncode != 0:
            raise RuntimeError(detail or "Julia IHW adapter failed")
        adjusted = _read_vector(output_prefix.with_suffix(".adj.txt"))
        weights = _read_vector(output_prefix.with_suffix(".weights.txt"))
        rejection_text = output_prefix.with_suffix(".rejections.txt").read_text(
            encoding="utf-8"
        )
    return FitResult(adjusted, weights, int(rejection_text.strip()))


def _read_vector(path: Path) -> np.ndarray:
    """Read one numeric vector written by the Julia adapter."""

    return np.atleast_1d(np.loadtxt(path, dtype=np.float64))


if __name__ == "__main__":
    raise SystemExit(adapter_main("julia_ihw", IMPLEMENTATION_VERSION, fit))
